amethyst_error/src/lib.rs

```diff
@@ -0,0 +1,422 @@
+    where
+        M: Into<Cow<'static, str>>,
+    {
+        /// Wrapper for string errors.
+        #[derive(Debug)]
+        struct StringError(Cow<'static, str>);
+
+        impl fmt::Display for StringError {
+            fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+                self.0.fmt(fmt)
+            }
+        }
+
+        impl error::Error for StringError {}
+
+        Self {
+            inner: Box::new(Inner {
+                source: None,
+                #[cfg(all(feature = "backtrace", feature = "std"))]
+                backtrace: self::internal::bt::new(),
+                error: Box::new(StringError(message.into())),
+            }),
+        }
+    }
+
+    /// Get backtrace.
+    pub fn backtrace(&self) -> Option<&Backtrace> {
+        #[cfg(all(feature = "backtrace", feature = "std"))]
+        {
+            self.backtrace.as_ref()
+        }
+
+        #[cfg(not(all(feature = "backtrace", feature = "std")))]
+        {
+            None
+        }
+    }
+
+    /// Get the source of the error.
+    ///
+    /// # Examples
+    ///
+    /// The only way to set the source is through [`ResultExt`](ResultExt) using
+    /// [`with_context`](ResultExt::with_context).
+    ///
+    /// ```rust
+    /// use amethyst_error::{Error, ResultExt};
+    /// use std::io;
+    ///
+    /// let e = io::Error::new(io::ErrorKind::Other, "wrapped");
+    /// let a = Error::new(e);
+    ///
+    /// let res = Result::Err::<(), Error>(a).with_context(|_| Error::from_string("top"));
+    /// let e = res.expect_err("no error");
+    ///
+    /// assert_eq!("top", e.to_string());
+    /// assert_eq!("wrapped", e.source().expect("no source").to_string());
+    /// ```
+    pub fn source(&self) -> Option<&Error> {
+        self.inner.source.as_ref().map(|e| &**e)
+    }
+
+    /// Iterate over all causes, including this one.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use amethyst_error::{Error, ResultExt};
+    ///
+    /// fn failing_function() -> Result<(), Error> {
+    ///     Err(Error::from_string("failing"))
+    /// }
+    ///
+    /// fn other_function() -> Result<(), Error> {
+    ///     Ok(failing_function().with_context(|_| Error::from_string("other"))?)
+    /// }
+    ///
+    /// let e = other_function().expect_err("no error");
+    ///
+    /// let messages = e.causes().map(|e| e.to_string()).collect::<Vec<_>>();
+    /// assert_eq!(vec!["other", "failing"], messages);
+    /// ```
+    pub fn causes(&self) -> Causes<'_> {
+        Causes {
+            current: Some(self),
+        }
+    }
+
+    /// Access the internal `std::error::Error` as a trait.
+    ///
+    /// This can be useful for integrating with systems that operate on `std::error::Error`.
+    ///
+    /// **Warning:** This erases most diagnostics in favor of returning only the top error.
+    /// `std::error::Error` is expanded further.
+    pub fn as_error(&self) -> &dyn error::Error {
```

Comment:
Similar to `std::error::Error` this should return `&dyn (Error + 'static)`

Reply:
Good call!